java/arcs/core/util/LruCacheMap.kt

```diff
@@ -55,10 +61,30 @@ open class LruCacheMap<K, V>(
         backingMap.put(key, it)
         return it
       } else {
-        onEvict?.invoke(key, it)
-        return null
+        scheduleEviction(key, it)
+        return pendingEviction.value.get(key)
       }
     }
     return value
   }
+
+  private fun runEvictor(key: K, value: V) {
+    evictor {
+      onEvict?.invoke(key, value)
+      pendingEviction.update { oldMap ->
+        oldMap.toMutableMap().also {
```

Comment:
Feel there would be potential exceptions happening at constructing oldMap into a new mutable map collection? How about considering a multimap plus synchronize() lock to maintain consistent map view?